### ISSUE TITLE
feeding back curr state to next state causing dynamic_rnn to crash if batch size is varying

### ISSUE THREAD

**Issue by amadupu** [tensorflow/tensorflow#20829]
Im feeding data to the graph with tf.train.batch (dynamic_pad=True,allow_smaller_final_batch=True).
also im feeding in the final state returned by tf.train.dynamic_rnn as the next state during the training process. (initial state configured to zeros)

it was working fine, however when end of epoch is reached, due to allow_smaller_final_batch=True above, it is trying to fetch smaller batch sizes. this seems to crash the program as the last state is of standard batch size(say 100) and current batch is lesser than standard batch size (eg: 90)

if i dont feed the last state, things are working fine as the state always initialized to zeros based on the current batch size. I could not find proper explanation for dynamic batch sizes (pls not not im not referring to variable sequence lenghts here)  assoiated with state feedback to ascertain if this is a bug or a coding issue

kindly help

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by drpngx** [comment]
Could you post the log for the crash?

**Comment by amadupu** [comment]
apologies for resonding late...please find the requested details as below

OS: Windows, 
IDE: Pycharm, 
Tensorflow Installed from: pip3
Tensorflow Version: 1.9.0


Steps to Reproduce:

- Created dynamic_rnn processing as follows
              outputs, self.final_state = tf.nn.dynamic_rnn(cell_fw,self.rnn_inputs,sequence_length=self.steps,
                                                                      initial_state=self.state_fw,time_major=False)

- Reading data from input pipelines as follows
       self.steps, self.xs , self.ys = tf.train.batch(tensors=decoder.dequeue(self.is_classifier), batch_size=50,
                                                                  **dynamic_pad=True**,
                                                                  **allow_smaller_final_batch=True**,name='batch_processor')

- During Training Phase Feeding the last state as follows
 _,loss,  accuracy, summary, final_state,     = self.sess.run([self.train_step,self.loss, self.accuracy, self.summary, self.final_state,],feed_dict)
feed_dict[self.state_fw] = final_state



- Error during the **last smaller batch**

_Dimensions of inputs should match: shape[0] = [20,512] vs. shape[1] = [50,512]
	 [[Node: rnn_layer/rnn/while/rnn/multi_rnn_cell/cell_0/LSTMCell/concat = ConcatV2[N=2, T=DT_FLOAT, Tidx=DT_INT32, _device="/job:localhost/replica:0/task:0/device:CPU:0"](rnn_layer/rnn/while/rnn/multi_rnn_cell/cell_0/dropout/mul, rnn_layer/rnn/while/Identity_4, rnn_layer/rnn/while/rnn/multi_rnn_cell/cell_0/LSTMCell/concat/axis)]]_

**Comment by drpngx** [comment]
It's not totally clear to me how you are calling this, but from what I gather, it looks like the last tensor is smaller. As you point out, you are using `allow_smaller_final_batch=True`, which is likely to produce this.

Why not set that to `False`? I'm not sure how it's connected to feeding back the last state.

Also, I don't exactly understand how you are going to process the continuous sequence unless you plan on truncating the gradient at each batch. (Just consider the initial state as a constant.)

**Comment by amadupu** [comment]
Yeah...currently im setting `allow_smaller_final_batch=False` to overcome this issue. As you have rightly pointed out, im having a trouble in feeding continous sequences. so ive truncated my data into shorter segments , and relying on state to carry forward to the next batch instead resetting the state. 

i realize this approach is not entirely correct as the batched data is no more continous as i expect to be.
im struggling on this point for a while may you can suggest a better way to do this

for eg:
let us say, continous data resembles 1,2,3........so on to 1000.
i have stored this data as follows
segment input data into smaller units of length 20

1,2,3,....20  (seg-1)
21,22,,,,40  (seg-2)
981,982,,,1000   (seg- 50)
all these segments are written to tfrecords as serialized sequence examples.

while training im reading the tfrecord file of fixed batch size (say 3)
RNN state S is of shape (batch size x cell size)

State     Batch -1                    New State
0          1,2,3..20                      S1  
0          21,22,...40                   S2
0          41,42...60                    S3

State     Batch -2                   New State
S1         61,62,...80                       S4  
S2          81,81...100                     S5
S3         101,102,...120                 S6


Question: is state S is correct representation of continous data? As S1,S2,S3 are representing states for different batch streams, im having a slight confusion if i need to ever use batching here or just always use batch-size of 1 with a single state 

Question: Are RNN models mostly suitable for smaller length sequences rather learning patterns in continous streams? though i dont believe so, but most of the documentation are talking around mini batches with state resets, im trying to get a basic idea where RNN are perfect fit for. 

Question: In what scenarios do we exactly carry forward state from one batch to other
